src/ngResource/resource.js

```diff
@@ -588,14 +657,23 @@ angular.module('ngResource', ['ng']).
               }
             });
 
+            if (!isInstanceCall && action.cancellable) {
+              var deferred = $q.defer();
+              httpConfig.timeout = deferred.promise;
+              value.$cancelRequest = deferred.resolve;
+            }
+
             if (hasBody) httpConfig.data = data;
             route.setUrlParams(httpConfig,
               extend({}, extractParams(data, action.params || {}), params),
               action.url);
 
-            var promise = $http(httpConfig).then(function(response) {
+            var promise = $http(httpConfig).finally(function() {
+              if (value.$cancelRequest) value.$cancelRequest = angular.noop;
```

Comment:
Note that we can't take care of it inside the `then` success callback, because there might be an error (in which case we'll never make it to the success callback).